feat(article): adaptive retry mode terraform (#160)
masterpointio/masterpoint.io: content/blog/2026-07-26-aws-adaptive-retry-mode-terraform-throttled-api-requests.md

resource "aws_iam_role" "app" {
  provider = aws.adaptive_requests
  for_each = local.roles
  # ...
}
```

Using [provider aliases with child modules](https://developer.hashicorp.com/terraform/language/modules/develop/providers#passing-providers-explicitly) is similar via `providers = { aws = aws.adaptive_requests }`. The blast radius is exactly the resources you chose, and it's all greppable when it's in code rather than an environment variable that applies to everything.

## Don't Apply It Everywhere

As per the AWS documentation, [adaptive mode is not recommended as a general default](https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html).

This is because there are costs. **Typical operations get slower** since delaying first attempts is the mechanism when there is throttling, so a Terraform/OpenTofu workspace that never gets throttled gains nothing and may still pay a latency cost after a transient blip. Also, that per-client rate limiter is **shared across operations**, so one throttle-prone API surface slows every call the client makes, including calls to operations that were never in trouble (hence the example above with Terraform provider aliases).

> The important nuance is that these costs only exist where requests were succeeding in the first place: for resources that are actually being throttled, "slower" is a non-issue, because the fast version wasn't going through anyway. An operation stuck cycling through `Rate exceeded` errors has no latency worth protecting, and a **paced request that succeeds beats multiple quick ones that bounces**. For the throttled resources, pacing is precisely what finally gets them through. So scope it to resources where throttling is the _expected_ failure mode. There, you're not giving up any speed because the requests via standard retries were already being throttled.

Author: done

author: Yangci Ou
slug: aws-adaptive-retry-mode-terraform-throttled-api-requests
date: 2026-07-26 # TODO: Update with correct date
# date_modified: 2026-xx-xx Be sure to use this if you've updated the post as this helps with SEO and index freshness
description: "Terraform's AWS provider retries throttled requests with exponential backoff, but each request backs off blind to the others. Adaptive retry mode adds a client-wide rate limiter that paces requests once throttling is detected to avoid being rate limited."
image: /img/updates/aws-adaptive-retry-mode-terraform/aws-adaptive-retry-mode.png
callout: <p>👋 <b>If you're ready to take your infrastructure to the next level, we're here to help. We love to work together with engineering teams to help them build well-documented, scalable, automated IaC that make their jobs easier. <a href='/contact'>Get in touch!</a></p>
---

<h2>Table of Contents</h2>

- [Standard Mode, where Every Request Backs Off Naively](#standard-mode-where-every-request-backs-off-naively)
- [Adaptive Mode, so the Client Paces Itself When Throttling is Detected](#adaptive-mode-so-the-client-paces-itself-when-throttling-is-detected)

>   - [Why Not Just Raise `max_retries`?](#why-not-just-raise-max_retries)

Author: ```suggestion
  - [Why Not Just Raise Max Retries?](#why-not-just-raise-max_retries)
```

- [Enable It in the TF Provider Block, Not the Environment](#enable-it-in-the-tf-provider-block-not-the-environment)
- [Don't Apply It Everywhere](#dont-apply-it-everywhere)

> **TL;DR:** By default, Terraform's AWS provider retries each throttled request with exponential backoff, but it has no idea what its other in-flight requests are doing. With concurrency, the client as a whole keeps slamming AWS' API. Using adaptive retry mode adds a client-wide rate limiter that paces requests once it detects throttling.

## Standard Mode, where Every Request Backs Off Naively

By default, the [Terraform AWS provider](https://registry.terraform.io/providers/hashicorp/aws/latest/docs) (via the AWS SDK underneath it) retries every throttled and rate-limited API request with exponential backoff and jitter. Each failed request waits, the wait roughly doubles on each attempt, and a random jitter is mixed in so retries don't synchronize into waves. If your Terraform or OpenTofu operation is drowning in `ThrottlingException` or `Rate exceeded` errors, it is not because Terraform forgot to back off.

> The default _standard_ retry mode backs off **per request**: each throttled call politely waits, doubles, waits again, with no awareness of what any other in-flight request is doing. Terraform is concurrent with a default `-parallelism` of 10 resource operations at once, and each operation is rarely a single API call (there's the create or update, the read-back to refresh state, and often a polling waiter checking whether the resource is ready). While one request sleeps through its backoff, nine siblings are firing fresh calls, and expired backoffs re-collide with new traffic. In an enterprise setting, where multiple operations are running at the same time from multiple systems, this compounds!

Author: ```suggestion
The default _standard_ retry mode backs off **per request**: each throttled call politely waits, doubles, waits again, with no awareness of what any other in-flight request is doing. Terraform is concurrent with a default `-parallelism` of 10 resource operations at once, and each operation is rarely a single API call (there's the create or update, the read-back to refresh state, and often a polling waiter checking whether the resource is ready). While one request sleeps through its backoff, nine other requests are firing fresh calls, and expired backoffs re-collide with new traffic. In an enterprise setting, where multiple operations are running at the same time from multiple systems, this compounds!
```

## Adaptive Mode, so the Client Paces Itself When Throttling is Detected

[Adaptive mode](https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html) keeps everything standard mode does and adds the missing piece, which is a client-side rate limiter that watches for throttling responses and dials the client's send rate up or down accordingly. When throttles appear, it cuts the rate and paces itself instead of blindly firing AWS API requests.

As the [AWS SDKs and Tools documentation on retry behavior](https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html) puts it, adaptive mode "can delay or block the _initial_ request, not just retries, when throttling is detected." That means the client's own sending is governed at the source.

[AWS recommends adaptive mode](https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html) for one specific shape of client.

> Your client targets a single resource (for example, one DynamoDB table) and you expect frequent throttling responses. This is common in automated workflows, batch processors, or AI workloads that call a single API operation at high volume.

> A big Terraform operation against a rate-capped service fits that description well. [IAM](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html) is the classic case, where you might have a monolithic Terraform root module with hundreds of roles and policies, [ALB listener rules](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-limits.html) in the thousands are another, where every rule is a Terraform resource whose refresh hammers the ELB Describe APIs, or [Route 53](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/DNSLimitations.html#limits-api-requests-route-53) is hard-capped at five API requests per second per account. **You should always request a Service Quotas increase first, but for limits like these, sometimes it gets denied or doesn't buy you enough headroom.**

Author: yes updated language

The provider block gives you both visibility and scoping. Since v5 of the Terraform AWS provider, [`retry_mode` is a natively supported provider argument](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#retry_mode-1):

```hcl
provider "aws" {
  region     = "us-east-1"
  retry_mode = "adaptive"
}
```

For further solution engineering to tailor the root module, you can go one step further and use [provider aliases](https://developer.hashicorp.com/terraform/language/block/provider#alias) so only certain resources prone to rate limits and throttling get adaptive mode:

```hcl
provider "aws" {
  region = "us-east-1" # default: standard retries
}

provider "aws" {
  alias      = "adaptive_requests"
  region     = "us-east-1"
  retry_mode = "adaptive" # paced client, scoped by alias
}

resource "aws_iam_role" "app" {
  provider = aws.adaptive_requests
  for_each = local.roles
  # ...
}
```

> Modules slot in the same way via `providers = { aws = aws.adaptive_requests }`. The blast radius is exactly the resources you chose, and it's all greppable.

Author: yes updated the language and linking to upstream docs. multiple ways of doing it.

---
visible: true
draft: false
title: "Using AWS Adaptive Retry Mode with Terraform for Throttled API Requests"
author: Yangci Ou
slug: aws-adaptive-retry-mode-terraform-throttled-api-requests

> date: 2026-07-26 # TODO: Update with correct date

Reviewer: ```suggestion
date: 2026-08-10
```

# date_modified: 2026-xx-xx Be sure to use this if you've updated the post as this helps with SEO and index freshness
description: "Terraform's AWS provider retries throttled requests with exponential backoff, but each request backs off blind to the others. Adaptive retry mode adds a client-wide rate limiter that paces requests once throttling is detected to avoid being rate limited."
image: /img/updates/aws-adaptive-retry-mode-terraform/aws-adaptive-retry-mode.png
---

<h2>Table of Contents</h2>

- [Standard Mode, Where Every Request Backs Off Naively](#standard-mode-where-every-request-backs-off-naively)
- [Adaptive Mode, so the Client Paces Itself When Throttling is Detected](#adaptive-mode-so-the-client-paces-itself-when-throttling-is-detected)
  - [Why Not Just Raise Max Retries?](#why-not-just-raise-max-retries)
  - [Why Not Just Lower Parallelism?](#why-not-just-lower-parallelism)
- [Enable It in the TF Provider Block, Not the Environment](#enable-it-in-the-tf-provider-block-not-the-environment)
- [Don't Apply It Everywhere](#dont-apply-it-everywhere)

> > **TL;DR:** By default, Terraform's AWS provider retries each throttled request with exponential backoff, but every request backs off on its own with no idea what the other in-flight requests are doing. With Terraform's parallelism, the client as a whole keeps slamming AWS' API, so the operation drags on and can burn through `max_retries`, erroring out or timing out entirely. Adaptive retry mode adds a client-wide rate limiter that paces requests once it detects throttling.

Reviewer 2: I’d describe it as an intro/overview/brief explanation. It covers the subject, but doesn’t offer a concise and actionable solution along the lines of: “you have problem X, which causes symptoms Y; so do Z to fix it.”
Or rephrase the last sentence to something like "Enable the Adaptive retry mode (`retry_mode = "adaptive"`) in your AWS provider config to fix that."

Reviewer: @gberenice good callout. Let's change this to "Intro" 👍

Reviewer: ```suggestion
## Intro

By default, Terraform's AWS provider retries each throttled request with exponential backoff, but every request backs off on its own with no idea what the other in-flight requests are doing. With Terraform's parallelism, the client as a whole keeps slamming AWS' API, so the operation drags on and can burn through `max_retries`, erroring out or timing out entirely. Adaptive retry mode adds a client-wide rate limiter that paces requests once it detects throttling. We'll dig into this functionality below. 
```

### Why Not Just Lower Parallelism?

The other knob is [`-parallelism`](https://developer.hashicorp.com/terraform/cli/commands/apply#parallelism-n), dropped from the default of 10 to 2 or 3 so fewer requests are in flight. That does reduce throttling, but it's a blunt instrument since parallelism is **global** while throttling almost never is. It's usually one or two API surfaces hitting their limits while everything else in the root module is perfectly fine, so **turning parallelism down slows every resource in the run to accommodate a small slice of it**, on every run, whether throttling shows up that day or not. Adaptive mode only kicks in once throttling is actually observed, and with provider aliases (more on that below) you can scope it to just the resources prone to rate limits while everything else runs at full speed.

## Enable It in the TF Provider Block, Not the Environment

You _can_ switch modes with the `AWS_RETRY_MODE` environment variable or the shared AWS config file, since the provider honors both.

<u>**We would recommend using the provider block, not the environment variable.**</u>

> An env var is invisible in code & reviews, lives outside version control, and applies to **everything** in that process and shell: every provider configuration in the run, and even the AWS CLI in the same CI job. That's the opposite of what we want, since adaptive should only be on where it's necessary, and reviewers & future readers should be able to _ easily see_ where it's on.

Author: fixed

  retry_mode = "adaptive" # paced client, scoped by alias
}

resource "aws_iam_role" "app" {
  provider = aws.adaptive_requests
  for_each = local.roles
  # ...
}
```

Using [provider aliases with child modules](https://developer.hashicorp.com/terraform/language/modules/develop/providers#passing-providers-explicitly) is similar via `providers = { aws = aws.adaptive_requests }`. The blast radius is exactly the resources you chose, and it's all greppable in code rather than an environment variable that's under the hood.

## Don't Apply It Everywhere

As per the AWS documentation, [adaptive mode is not recommended as a general default](https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html).

This is because there are costs. **Typical operations get slower**, since delaying first attempts is the mechanism itself, so a Terraform or OpenTofu workspace that never gets throttled gains nothing and may still pay a latency cost after a transient blip. And that per-client rate limiter is **shared across operations**, so one throttle-prone API surface slows every call the client makes, including ones that were never in trouble (hence the provider aliases above).

> The important nuance is that these costs only exist where requests were succeeding in the first place. For resources that are actually being throttled, "slower" is a non-issue because it wasn't going through anyway and cycling through `Rate exceeded` errors has no latency worth protecting. A **paced request that succeeds beats multiple quick ones that bounce**, so scope it to resources where throttling is expected, where you give up no real speed because standard retries were already getting throttled.

Reviewer 2: While the message is pretty straightforward, I still feel like a short summary is missing. Without it, it feels like we drop a reader in the middle of a conversation.
Maybe something like:
>The practical rule is simple: keep standard mode by default, and use adaptive mode where throttling is expected and repeatable. A paced request that succeeds is better than a burst of fast requests that AWS rejects.

Reviewer: Agreed. The ending was not succinct enough. I like yours. Let's update. 

```suggestion
We have a practical rule at Masterpoint that is simple: keep the standard mode by default, and use adaptive retry mode where throttling occurs or is expected (Route53, we're looking at you). A paced request that succeeds is better than a burst of fast requests that AWS rejects.
```

Reviewer 2: >Route53, we're looking at you

Haha, I like it!

Reviewer 2: @Gowiem I didn't fully remove the paragraph "The important nuance is that...", but I edited it to preserve the reasoning.

> The important nuance is that these costs only exist where requests were succeeding in the first place. For resources that are actually being throttled, "slower" is a non-issue because the fast version wasn't going through anyway: an operation stuck cycling through `Rate exceeded` errors has no latency worth protecting, and a **paced request that succeeds beats multiple quick ones that bounce**. So scope it to resources where throttling is expected, where you give up no real speed because standard retries were already getting throttled.

Author: fixed
